from core.integrations.source_metadata_extractors.bash_metadata_extractor import BashSourceMetadataExtractor
from core.integrations.source_metadata_extractors.github_metadata_extractor import GithubSourceMetadataExtractor
from core.integrations.source_metadata_extractors.jenkins_metadata_extractor import JenkinsSourceMetadataExtractor
from core.integrations.source_metadata_extractors.kubectl_metadata_extractor import KubernetesMetadataExtractor
from core.protos.base_pb2 import Source
from core.integrations.source_metadata_extractors.azure_metadata_extractor import AzureConnectorMetadataExtractor
from core.integrations.source_metadata_extractors.clickhouse_metadata_extractor import ClickhouseSourceMetadataExtractor
from core.integrations.source_metadata_extractors.cloudwatch_metadata_extractor import CloudwatchSourceMetadataExtractor
from core.integrations.source_metadata_extractors.datadog_metadata_extractor import DatadogSourceMetadataExtractor
from core.integrations.source_metadata_extractors.eks_metadata_extractor import EksSourceMetadataExtractor
from core.integrations.source_metadata_extractors.elastic_search_metadata_extractor import \
    ElasticSearchSourceMetadataExtractor
from core.integrations.source_metadata_extractors.gke_metadata_extractor import GkeSourceMetadataExtractor
from core.integrations.source_metadata_extractors.grafana_metadata_extractor import GrafanaSourceMetadataExtractor
from core.integrations.source_metadata_extractors.newrelic_metadata_extractor import NewrelicSourceMetadataExtractor
from core.integrations.source_metadata_extractors.open_search_metadata_extractor import OpenSearchSourceMetadataExtractor
from core.integrations.source_metadata_extractors.jira_metadata_extractor import JiraSourceMetadataExtractor
from core.integrations.source_metadata_extractors.argocd_metadata_extractor import ArgoCDSourceMetadataExtractor
from core.integrations.source_metadata_extractor import SourceMetadataExtractor
from core.integrations.source_metadata_extractors.posthog_metadata_extractor import PosthogSourceMetadataExtractor
from core.integrations.source_metadata_extractors.signoz_metadata_extractor import SignozSourceMetadataExtractor
from core.integrations.source_metadata_extractors.mongodb_metadata_extractor import MongoDBSourceMetadataExtractor
from core.integrations.source_metadata_extractors.postgres_metadata_extractor import PostgresSourceMetadataExtractor
from core.integrations.source_metadata_extractors.sql_database_connection_metadata_extractor import SqlDatabaseConnectionMetadataExtractor
from core.integrations.source_metadata_extractors.gcm_metadata_extractor import GcmSourceMetadataExtractor

class SourceMetadataExtractorFacade:

    def __init__(self):
        self._map = {}

    def register(self, source: Source, metadata_extractor: SourceMetadataExtractor.__class__):
        self._map[source] = metadata_extractor

    def get_connector_metadata_extractor_class(self, connector_type: Source):
        if connector_type not in self._map:
            raise ValueError(f'No metadata extractor found for connector type: {connector_type}')
        return self._map[connector_type]


source_metadata_extractor_facade = SourceMetadataExtractorFacade()
source_metadata_extractor_facade.register(Source.DATADOG, DatadogSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.DATADOG_OAUTH, DatadogSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.CLOUDWATCH, CloudwatchSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.NEW_RELIC, NewrelicSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.GRAFANA, GrafanaSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.GKE, GkeSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.KUBERNETES, KubernetesMetadataExtractor)
source_metadata_extractor_facade.register(Source.CLICKHOUSE, ClickhouseSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.EKS, EksSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.AZURE, AzureConnectorMetadataExtractor)
source_metadata_extractor_facade.register(Source.ELASTIC_SEARCH, ElasticSearchSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.OPEN_SEARCH, OpenSearchSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.GITHUB, GithubSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.JIRA_CLOUD, JiraSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.ARGOCD, ArgoCDSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.BASH, BashSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.JENKINS, JenkinsSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.POSTHOG, PosthogSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.SIGNOZ, SignozSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.MONGODB, MongoDBSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.POSTGRES, PostgresSourceMetadataExtractor)
source_metadata_extractor_facade.register(Source.SQL_DATABASE_CONNECTION, SqlDatabaseConnectionMetadataExtractor)
source_metadata_extractor_facade.register(Source.GCM, GcmSourceMetadataExtractor)